rgw/simplefile: Sqlite users using sqlite_orm library

0xavi0 opened the pull request:
Access SQLite database in a more modern C++ fashion.

This this a preliminary version (prove of concept mainly) and includes the header for sqlite_orm. It should be removed in future versions and it's only intended as a demo to see if this is feasible or not. We also should double check the license.

The database is stored in the folder defined by the `rgw_simplefile_data_path` variable.

The database can be accessed in 2 ways.

### Using direct calls to sqlite_orm's storage ###
This example from unit tests shows how to access the storage class and do your own sqlite_orm calls.
The main inconvenient is that **so far** I could not find how to use a RGWUser class directly with sqlite_orm and you need to convert the DBUser to RGW's one.
I think the main problem is that sqlite_orm doesn't like nested types and things like `rgwuser.uinfo.user_id.id`, but I can investigate more if we find this library useful.


For other use cases (like metadata in which we control the types we use) I don't think we need to implement conversions.

Also, sqlite_orm is template based and we can specialise encoders and decoders, so direct conversion to complex types like `std::map< std::string, RGWAccessKeys >` is doable. (I've tested it and works fine, but I thought specialising so many templates is more painful to maintain).
 
```c++
TEST_F(TestSimpleFileSQLiteUsers, UseStorage) {
  auto ceph_context = std::make_shared<CephContext>(CEPH_ENTITY_TYPE_CLIENT);
  ceph_context->_conf.set_val("rgw_simplefile_data_path", getTestDir());

  SQLiteUsers db_users(ceph_context.get());
  auto storage = db_users.getStorage();

  DBUser db_user;
  db_user.UserID = "test_storage";

  // we have to use replace because the primary key of rgw_user is a string
  storage.replace(db_user);

  auto user = storage.get_pointer<DBUser>("test_storage");

  ASSERT_NE(user, nullptr);
  ASSERT_EQ(user->UserID, "test_storage");

  // convert the DBUser to RGWUser (blobs are decoded here)
  auto rgw_user = getRGWUser(*user);
  ASSERT_EQ(rgw_user.uinfo.user_id.id, user->UserID);

  // creates a RGWUser for testing (id = test1, email = test1@test.com, etc..)
  auto rgw_user_2 = createTestUser("1");

  // convert to DBUser (blobs are encoded here)
  auto db_user_2 = getDBUser(rgw_user_2);

  // we have to use replace because the primary key of rgw_user is a string
  storage.replace(db_user_2);

  // now use the SqliteUsers method, so user is already converted
  auto ret_user = db_users.getUser("test1");
  ASSERT_TRUE(ret_user.has_value());
  compareUsers(rgw_user_2, *ret_user);
}
```
### Using the helpers provided by SQLiteUsers class ###
Conversions are done automatically

Example of use in RGW's code
```c++
int SimpleFileStore::get_user_by_access_key(const DoutPrefixProvider *dpp,                                    
                                            const std::string &key,                                           
                                            optional_yield y,                                                 
                                            std::unique_ptr<User> *user) {                                                                                                                                                                                                      
    int err = 0;                                                                                                
    rgw::sal::simplefile::sqlite::SQLiteUsers sqlite_users(dpp->get_cct());                                     
    auto db_user = sqlite_users.getUserByAccessKey(key);                                                        
    if (db_user) {                                                                                              
      user->reset(new SimpleFileUser(db_user->uinfo, this));                                                    
    } else {                                                                                                    
      ldpp_dout(dpp, 10) << __func__ << ": User not found" << dendl;                                            
      err = -1;       // or whatever code, I'm just an example                                                                                          
    }                                                                                                           
    return err;                                                                                                 
  }    
```

The interface is now pretty simple and returns `std::optional` for methods that expect a single object.
Queries by email and keys are also returning the first entry found in the database.

I guess those columns should be flagged as unique, so the query returns 1 single user (or none).

Regarding sqlite_orm. I had to disable optimisations in the header here... because gcc was killed by the OOM killer.

Signed-off-by: 0xavi0 <xavi.garcia@suse.com>

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [X] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [X] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [X] No doc update is appropriate
- Tests (select at least one)
  - [X] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

jecluis commented:
I think this looks great :)

However, I also think we should have the `sqlite_orm` library being installed on the system as a dependency, rather than including it in the project.

0xavi0 commented:
> However, I also think we should have the `sqlite_orm` library being installed on the system as a dependency, rather than including it in the project.
Yup, totally agree. 
It's there so we can have a quick look to see if it's a good option. I think it is a good improvement compared to the DBStore approach.

I also will investigate what's going on with the optimisation issues I had when building

jecluis reviewed (commented):
@0xavi0 looks good, but two comments:

1. your commit message is

```
rgw/simplefile: Sqlite users using sqlite_orm library
Signed-off-by: 0xavi0 <xavi.garcia@suse.com>
```
and it should read your name instead of github user.

2. Would be amazing if you could add some description to the code on how the data is mapped to the database, and how we can expand this for other things.

0xavi0 commented:
I've changed the name in the commit message, added a few comments about `sqlite_orm` and integrated with the package from OBS instead of adding the whole `sqlite_orm` file to the repo.

jecluis commented on src/rgw/rgw_sal_simplefile.h:
```diff
@@ -246,7 +245,7 @@ class SimpleFileStore : public Store {
     return 0;
   }
-
+  
```
unnecessary blank space

jecluis commented on src/rgw/store/simplefile/sqlite/sqlite_orm.h:
```diff
@@ -0,0 +1,10 @@
+#pragma once
+
+// TODO investigate with optimisations are causing gcc to get killed
```
s/with/which/

jecluis commented on src/rgw/store/simplefile/user.cc:
```diff
@@ -83,9 +84,19 @@ int SimpleFileUser::load_user(const DoutPrefixProvider *dpp, optional_yield y) {
-  return -ENOTSUP;
+  // /** Store this User to the backing store */ ldpp_dout(dpp, 10)
+  //     << __func__ << ": TODO" << dendl;
+  // return -ENOTSUP;
```
might just as well drop this

jecluis reviewed (commented):
looks good to me, few nits, but good to go imo

0xavi0 commented:
thanks, @jecluis!
I think those 3 should be okay now.

jecluis reviewed (approved):
have not tested, but I'll assume it's working ;)

giubacc commented:
I have built/run the s3gw and successfully executed the smoke tests.
The code looks good for me.
As said via slack, I would then add in the Dockerfile the dependency vs sqlliteorm.

giubacc reviewed (commented):
LGTM
